common/model/bundle.c

```diff
@@ -236,6 +257,73 @@ void bundle_reset_indexes(bundle_transactions_t *const bundle) {
   }
 }
 
+void bundle_set_messages(bundle_transactions_t *bundle, signature_fragments_t *messages) {
+  iota_transaction_t *tx = NULL;
+  size_t index = 0;
+  tryte_t **msg = NULL;
+  size_t msg_len = 0;
+  tryte_t trytes_buff[NUM_TRYTES_SIGNATURE];
+
+  BUNDLE_FOREACH(bundle, tx) {
+    msg = signature_fragments_at(messages, index);
+    if (msg == NULL) {
+      break;
+    }
+    msg_len = strlen((char *)*msg);
+    memcpy(trytes_buff, *msg, msg_len);
+    memset(trytes_buff + msg_len, '9', NUM_TRYTES_SIGNATURE - msg_len);
+
+    // trytes to flex_trits
+    flex_trits_from_trytes(tx->data.signature_or_message, NUM_TRITS_SIGNATURE, trytes_buff, NUM_TRYTES_SIGNATURE,
+                           NUM_TRYTES_SIGNATURE);
+    index++;
+  }
+}
+
+retcode_t bundle_sign(bundle_transactions_t *const bundle, flex_trit_t const *const seed, inputs_t const *const inputs,
+                      Kerl *const kerl) {
+  iota_transaction_t *tx = NULL;
+  input_t *input = NULL;
+  size_t curr_index = 0;
+  flex_trit_t *signed_signature = NULL;
+
+  bundle_reset_indexes(bundle);
+  bundle_finalize(bundle, kerl);
+
+  // find the inputs and get the corresponding private key and calculate the signature fragment
+  BUNDLE_FOREACH(bundle, tx) {
+    if (transaction_value(tx) < 0) {  // input transactions
+      INPUTS_FOREACH(inputs->input_array, input) {
+        if (memcmp(input->address, transaction_address(tx), FLEX_TRIT_SIZE_243) == 0) {
+          if (curr_index > transaction_current_index(tx)) {
+            continue;
+          }
+          signed_signature =
+              iota_sign_signature_gen_flex_trits(seed, input->key_index, input->security, transaction_bundle(tx));
+          if (signed_signature) {
+            // for each security level add signature
+            for (int i = 0; i < input->security; i++) {
+              tx = bundle_at(bundle, curr_index + i);
+              memcpy(tx->data.signature_or_message, signed_signature + (i * NUM_FLEX_TRITS_SIGNATURE),
+                     NUM_FLEX_TRITS_MESSAGE);
+              tx->loaded_columns_mask.data |= MASK_DATA_SIG_OR_MSG;
```

Comment:
Do it only once after the loop ?

Reply:
for what? it sets the flag to tx after filling.

Reply:
But why to set it as many time as the security when you can set it once after ?

Reply:
no thing about security level, treat it as `transaction_set_message`

Reply:
Thanks, I was missing the fact that `tx` is not the same as in the `BUNDLE_FOREACH ` loop